doc/transfig.1
.TH TRANSFIG 1 "Version 3.2.4 November 2002"
.SH NAME
transfig \- creates a makefile for portable LaTeX figures
.SH SYNOPSIS
.B transfig
[ -L \fIlanguage\fR ]
[ -M \fImakefile\fR ]
[ -T \fItexfile\fR ]
[ -I \fImacrofile\fR ]
[ -V ] 
[ \fIcontrol_specs\fR figfile ... ]
.SH DESCRIPTION
.I Transfig
creates a makefile to translate figures described in Fig code
or PIC into a specified LaTeX graphics \fIlanguage\fR.
PIC files are identified by the suffix ".pic";
Fig files can be specified either with or without the suffix ".fig" .
Transfig also creates a TeX macro file appropriate to the target language.
.SH OPTIONS
The \fIlanguage\fR specifier must be one of
.BR box,
.BR cgm,
.BR eepic,
.BR eepicemu,
.BR emf,
.BR epic,
.BR eps,
.BR ge,
.BR gif,
.BR ibmgl,
.BR jpeg,
.BR latex,
.BR map,
.BR mf,
.BR mmp,
.BR mp,
.BR pcx,
.BR pdf,
.BR pdftex,
.BR pdftex_t,
.BR pic,
.BR pictex,
.BR png,
.BR ppm,
.BR ps,
.BR psfig,
.BR pstex,
.BR pstex_t,
.BR ptk,
.BR sld,
.BR textyl, 
.BR tiff,
.BR tk,
.BR tpic,
.BR xbm,
.BR xpm
or
.BR null .
The specifier
.B psfig
is like the 
.B ps
specifier, except that it
make use of the psfig macro package.
The specifier
.B pstex
is like
.B psfig,
except that it uses LaTeX to process formatted (special) text.
The specifier 
.B null
results in the figures being replaced by empty boxes of the
appropriate size.
If no language is specified,
.B epic
is the default.
.PP
The names of the makefile and TeX macro file can be set with the
-M and -T options.
The default names are "Makefile" and "transfig.tex", respectively.
If there is already an existing \fImakefile\fR in the directory, transfig
first renames it to \fImakefile~\fR.  The same holds for any existing TeX macro file.
If the -I option is specified, then a command to read in \fImacrofile\fR is
	inserted into the TeX macro file.
If the -V option is specified, the program version number is printed only.
.PP
Control specifiers are 
.TP
\fB\-m\fI mag\fR
set the magnification at which the figure is rendered (default 1.0).
.TP
\fB\-f\fI font\fR
set the default font family used for text objects (default "\fBrm\fR").
.TP
\fB\-s \fIfontsize\fR
set the default font size (in points) for text objects (default 11*\fImag\fR).
.TP
\fB\-o \fIoption\fR
pass the specified option string through to fig2dev.
.PP
If there is no scaling, the default font is an eleven point roman font.
Language and control specifiers may occur more than once;
they affect the translation of files which appear to their
right in the argument list. 
If no files are specified, then "*.fig" is assumed.
.SH "SEE ALSO"
fig(l),
fig2dev(l),
pic(l),
pic2fig(l),
xfig(l)
.SH COPYRIGHT
Copyright (C) 1985 Supoj Sutantavibul
.br
Copyright (C) 1991 Micah Beck
.br
Parts Copyright (C) 1991-2002 Brian Smith
.LP
THE AUTHORS DISCLAIM ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO
EVENT SHALL THE AUTHORS BE LIABLE FOR ANY SPECIAL, INDIRECT OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
.LP
Any party obtaining a copy of these files is granted, free of charge, a
full and unrestricted irrevocable, world-wide, paid up, royalty-free,
nonexclusive right and license to deal in this software and
documentation files (the "Software"), including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons who receive
copies from any such party to do so, with the only requirement being
that this copyright notice remain intact.

.SH AUTHOR
Micah Beck


CHANGES
Changes to Transfig/fig2dev
===============================================================================
Version 3.2
===============================================================================

Patchlevel 5e (August 2013)

BUGS FIXED:
	o HTML map output was limited to 100 links.
	  Fix by Jan van Dijk
	o Updated for compatibility to PNG 1.5
	  From Matthias Scheler
	o Was adding "showpage" command when producing bitmap formats from intermediate EPS.
	  This produced "illegal" PNG and JPEG files with extra, blank image.
	o Maximum width of included image in PS/EPS output increased from 8192 to 16384
	o Precision of some PIC objects increased from %.2f to %.3f
	o Double close of output file when ghostscript fails. Original bug report from
	  https://bugzilla.redhat.com/728825

-------------------------------------
Patchlevel 5d (July 23, 2010)

BUGS FIXED:
	o reduces some compilation warnings (fix from Debian)
	o made PostScript output DSC 3.0 compliant to work with CUPS 
	  patch from Ian Dall (see https://bugzilla.redhat.com/558380)
	o Changed STOCK_LAST from 17 to 19 in fig2dev/dev/genemf.c
	o Questionable copy of one data type to another in genemf.c
	o Uninitialized linew_spec in genepic.c
	o Changed definition of command for short slanted lines in genlatex.c
	o Added check for existance of arrows in SVG line generator
	o Removed %%Page: directive from included JPEG files in PostScript/EPS output
	o bound.c was passing pointers to int instead of double to arc_tangent
	o put_msg function uses proper varargs now
	o Option for debug comments in GBX output is "yes" or "no", not "on" or "off"
	o Added help (-h)  info for GBX output
	o In version 1.4 of the PNG library dither was removed so fig2dev now uses
	  quantize when importing PNG images with palettes
	o put_msg was declared twice (fig2dev/dev/genemf.c)
	o Added xlink namespace for images in SVG export
	o SVG output for imported images didn't form href link properly
	o PIC floating point numbers were only 2 decimal places - now 3 (Debian bug #611948)
	  Fix thanks to Roger Leigh (Debian Project)
	o PIC output text had leading blanks (Debian bug #611977)
	  Fix thanks to Roger Leigh (Debian Project)
	o LaTeX commands for small and tiny dots missing "." (e.g. \small.)
	  from Thomas Loimer

NEW FEATURES:
	o New PSTricks driver from Gene Ressler (see man fig2dev for info)

-------------------------------------
Patchlevel 5c (Jan 4, 2010)

BUGS FIXED:
	o transfig command was hardwired for "tex" for "make all" directive and
	  removing files with "make clean" directive in creating Makefile

-------------------------------------
Patchlevel 5b (Jun 2, 2009)

NEW FEATURES:
	o -a option added to PostScript and PICTeX languages to NOT include user's
	  login name in output
	o Gerber (RS-247-X for CAD drawings for printed circuits) export language from Edward Grace

-------------------------------------
Patchlevel 5a (Jun 1, 2009)

NEW FEATURES:
	o 22 new arrowhead types
	o HTML map now produces reference to .png file instead of .gif

BUGS FIXED:
	o CreationDate in pictex output was either wrong or caused segfault
	o bugs fixed in SVG Driver:
	  - blue component of shaded colors was always zero
	  - line protruding beyond arrowhead on long arrows
	  - wrong position of back arrowhead on double-headed arrows
	  - fill patterns either missing or using incorrect linewidth
	  - lowercase greek phi did not match its X11 counterpart
	  from Martin Kroeker
	o dubious printf(j++,j++) in MP driver
	o missing #includes in MP driver
	o added !defined(__FreeBSD) to fig2dev.h def for sys_nerr and errno
	  From Eric Scott
	o added -quiet options to giftopnm and ppmtopcs in reading GIF files
	  From Eric Scott
	o Typo in LaTeX driver when generating arc-box. "iut" should be "put"
	o extraneous stroke:black in svg header removed along with stroke color
	  for text, since the text fill color does the whole job
	o fig2dev produces more correctly structured PostScript.  Files get printed via
	  CUPS on PostScript printers.  When viewing a .ps-file in gv you can go to the
	  last page and back and to the last page again without getting PostScript errors.
	  From Ronald Lembcke
	o Renamed the macro \SetFigFont to \SetFigFontNFSS if NFSS is #defined to alleviate
	  some problems. From Roland Rosenfeld.
	o Grid in metric mode was incorrectly scaled
	o Removes any %EOF or %%EOF from imported EPS pictures before exporting.  
	  Also doesn't add %EOF when importing JPEG file now.
	o Missing call to get local time before converting to string in genpictex.c

-------------------------------------
Patchlevel 5 (Feb 27, 2007)

NEW FEATURES:
	o SVG driver adds color and pattern filling to arcs
	  from Martin Kroeker

BUGS FIXED:
	o bug which made arrowheads too short fixed
	o SVG driver corrects the ugly arrowheads on stick arrows
	  from Martin Kroeker
	o EMF driver would crash when fed stdin
	  from Martin Kroeker
	o Unhandled case for linetypes in latex driver
	o CUPS didn't like the order of some of the Postscript header
	  from Ian Dall

-------------------------------------
Patchlevel 5-alpha7 (Oct, 27, 2005)

NEW FEATURES
	o Updated SVG driver from Martin Kroeker was missing from alpha6
	o Another SVG update: Use a font-family list of "Times,Symbol" for symbol
	  characters - the Times fontface does not contain all elements of the
	  Symbol font on all platforms
	
BUGS FIXED:
	o Typo in font name (was cmit10, should be cmti10) in MetaPost driver
	  This bug fix was omitted from alpha6

Patchlevel 5-alpha6 (Oct, 12, 2005)

NEW FEATURES:
	o Added -dPDFSETTINGS=/prepress option to ghostscript when generating
	  PDF output to improve quality of bitmap graphics
	o Fill patterns now use PostScript Level 2 built-in pattern commands
	  for Postscript/EPS/PDF and bitmap output
	  From Konstantin Shemyak
	o Updated SVG driver from Martin Kroeker
	o New -K option to size bounding box to whole figure when exporting
	  only certain depths
	  From Dirk Osswald

BUGS FIXED:

transfig/transfig.c
/*
 * TransFig: Facility for Translating Fig code
 * Copyright (c) 1991 by Micah Beck
 * Parts Copyright (c) 1985-1988 by Supoj Sutanthavibul
 *
 * Any party obtaining a copy of these files is granted, free of charge, a
 * full and unrestricted irrevocable, world-wide, paid up, royalty-free,
 * nonexclusive right and license to deal in this software and
 * documentation files (the "Software"), including without limitation the
 * rights to use, copy, modify, merge, publish and/or distribute copies of
 * the Software, and to permit persons who receive copies from any such 
 * party to do so, with the only requirement being that this copyright 
 * notice remain intact.
 *
 */

/*
 * transfig: 	figure translation setup program
 *		creates TeX macro file and makefile
 *
 * usage: transfig <option> ... [[<flag> ... ] [<figure>] ... ] ...
 *
 * where:	<option> = -L <language> | -M <makefile> | -T <texfile>
 *		<flag>	 = -f <font> | -s <size> | -m <scale>
 */

#include <stdio.h>
#include <stdlib.h>
#include "patchlevel.h"
#include "transfig.h"


argument *parse_arg(), *arglist = NULL, *lastarg = NULL;
char *strip();

char *mkfile = "Makefile";
char *txfile = "transfig.tex";
char *input = "";
int  altfonts = 0;

/* output languages */
/* Warning - if this list is changed the enum language list in transfig.h must be changed too */

char *lname[] = {
		"box",
		"cgm",
		"eepic",
		"eepicemu",
                "emf",
		"epic",
		"eps",
		"gbx",
		"gif",
		"ibmgl",
		"dxf",
		"jpeg",
		"latex",
		"map",
		"mf",
                "mmp",
                "mp",
		"pcx",
		"pdf",
		"pdftex",
		"pic",
		"pictex",
		"png",
		"ppm",
		"ps",
		"psfig",
		"pstex",
		"ptk",
		"sld",
		"svg",
		"textyl",
		"tiff",
		"tk",
		"tpic",
		"xbm",
		"xpm"};

/* enum input {apg, fig, pic, ps}; */
char *iname[] = {
	"apg",
	"fig",
  	"pic",
	"ps",
	"eps"};
 
main(argc, argv)
int argc;
char *argv[];
{
  FILE *mk, *tx;
  enum language tolang = epic;
  argument *a;
  char *cp; 
  char *arg_f = NULL, *arg_s = NULL, *arg_m = NULL, *arg_o = NULL, *argbuf;

  for ( optind = 1; optind < argc; optind++ ) {
    cp = argv[optind];
    if (*cp == '-') {
  	if (!cp[1]) {
		fprintf(stderr, "transfig: bad option format '-'\n");
		exit(1);
	}
	if (cp[1] == 'V') {
		fprintf(stderr, "TransFig Version %s Patchlevel %s\n",
							VERSION, PATCHLEVEL);
		exit(0);
	} else if (cp[1] == 'h') {
		fprintf(stderr,"usage: transfig <option> ... [[<flag> ... ] [<figure>] ... ] ...\n");
		fprintf(stderr,"where:	<option> = -L <language> | -M <makefile> | -T <texfile>\n");
		fprintf(stderr,"	<flag>	 = -f <font> | -s <size> | -m <scale>\n");
		exit(0);
	}

	if (cp[2]) {
		optarg = &cp[2];
	} else {
	    if (cp[1] != 'a') {
		optind += 1;
		if (optind == argc) {
		    fprintf(stderr, "transfig: no value for '%c' arg\n", cp[1]);
		    exit(1);
		}
		optarg = argv[optind];
	    }
	}
 	switch (cp[1]) {

	    case 'I':
		input = optarg;
		break;

  	    case 'L':
		tolang = str2lang(optarg);
		break;
  	    case 'M':
		mkfile = optarg;
		break;
  	    case 'T':
		txfile = optarg;
		break;
	    case 'a':
		altfonts = 1;
		break;
	    case 'f':
		arg_f = optarg;
		break;
	    case 's':
		arg_s = optarg;	
		break;
	    case 'm':
		arg_m = optarg;	
		break;

	    case 'o':
		arg_o = optarg;
		break;

  	default:
		fprintf(stderr, "transfig: illegal option -- '%c'\n", cp[1]);
		exit(1);
  	}
    } else
    {
	a = parse_arg(tolang, arg_f, arg_s, arg_m, arg_o, argv[optind]);

	if ( !lastarg )
		arglist = a;
	else
		lastarg->next = a; 
	lastarg = a;
    }
  }

  /* no files specified -> all files */
  if (!arglist)
  {
	argbuf = sysls();
	while (cp = strchr(argbuf, '\n'))
	{
		*cp = '\0';
		a = parse_arg(tolang, arg_f, arg_s, arg_m, arg_o, argbuf);
		if ( !lastarg )
			arglist = a;
		else
			lastarg->next = a; 
		lastarg = a;
		argbuf = cp+1;
	}
  }

  /* remove any backup and move existing texfile to texfile~ */
  sysmv(txfile);
  if (strcmp(txfile,"-")==0)
	tx = stdout;
  else if ((tx = fopen(txfile, "w")) == NULL) {
	fprintf(stderr,"transfig: can't open Texfile '%s'\n",txfile);
	exit(1);
  }
  texfile(tx, input, altfonts, arglist);

  /* remove any backup and move existing makefile to makefile~ */
  sysmv(mkfile);
  if (strcmp(mkfile,"-")==0)
	mk = stdout;
  else if ((mk = fopen(mkfile, "w")) == NULL) {
	fprintf(stderr,"transfig: can't open Makefile '%s'\n",mkfile);
	exit(1);
  }
  makefile(mk, altfonts, arglist);
  exit(0);
}

enum language str2lang(s)
char *s;
{
  int i;

  /* aliases */
  if (!strcmp(s, "pic")) return tpic;
  if (!strcmp(s, "postscript")) return ps;
  if (!strcmp(s, "latexps")) return pstex;
  if (!strcmp(s, "null")) return box;

  /* real names*/
  for (i = 0; i <= (int)MAXLANG; i++)
	if (!strcmp(lname[i], s)) return (enum language)i;

  /* other strings */
  fprintf(stderr, "Unknown output language \"%s\"\n", s);
  exit(1);
}

argument *parse_arg(tolang, arg_f, arg_s, arg_m, arg_o, arg)
enum language tolang;
char *arg_f, *arg_s, *arg_m, *arg_o, *arg;
{
  argument *a;

  a = (argument *)malloc(sizeof(argument));
  a->f = arg_f;
  a->s = arg_s;
  a->m = arg_m;
  a->o = arg_o;
  a->next = NULL;
  a->tofig = NULL;
  a->topic = NULL;
  a->tops = NULL;
  a->tolang = tolang;
  
  /* PIC */
  if (strip(arg, ".pic"))
  {
  	a->name = mksuff(arg, "");
  	a->type = i_pic;
	a->tofig = PIC2FIG;
	return a;
  }

  /* PS format */
  if (strip(arg, ".ps"))
  {
  	a->name = mksuff(arg, "");
  	a->type = i_ps;
 	return a;
  }

  /* EPS format */
  if (strip(arg, ".eps"))
  {
  	a->name = mksuff(arg, "");
  	a->type = i_eps;
 	return a;
  }

  /* ApGraph format */
  if (strip(arg, ".apg"))
  {
  	a->name = mksuff(arg, "");
  	a->type = i_apg;
	a->tofig = APG2FIG;
 	return a;
  }

  /* Fig format */
  strip(arg, ".fig");
  a->name = mksuff(arg, "");
  a->type = i_fig;
  return a;
}
